File: 74. Search a 2D Matrix/74. Search a 2D Matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])

        low = 0
        high = m * n - 1

        while low <= high:
            mid = (low + high) // 2
            mid_i = mid // n
            mid_j = mid % n
            mid_matrix = matrix[mid_i][mid_j]

            if mid_matrix == target:
                return True
            elif mid_matrix > target:
                high = mid - 1
            else:
                low = mid + 1

        return False

File: 74. Search a 2D Matrix/74. Search a 2D Matrix.java
class Solution {

  public boolean searchMatrix(int[][] matrix, int target) {
    int m = matrix.length;
    int n = matrix[0].length;

    int low = 0;
    int high = m * n - 1;

    while (low <= high) {
      int mid = (low + high) / 2;
      int midI = mid / n;
      int midJ = mid % n;
      int midMatrix = matrix[midI][midJ];

      if (midMatrix == target) {
        return true;
      } else if (midMatrix > target) {
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    }

    return false;
  }
}
